editor/components/inserter/menu.js

```diff
@@ -2,20 +2,83 @@
  * External dependencies
  */
 import { connect } from 'react-redux';
+import { Observable } from 'rxjs/Observable';
+import 'rxjs/add/observable/fromEvent';
+import 'rxjs/add/operator/filter';
 
 /**
  * Internal dependencies
  */
 import './style.scss';
 import Dashicon from 'components/dashicon';
+import { focusNextFocusableRef, focusPreviousFocusableRef, listVisibleBlocksByCategory } from './helpers';
 
 class InserterMenu extends wp.element.Component {
 	constructor() {
 		super( ...arguments );
+		this.blockTypes = wp.blocks.getBlocks();
+		this.categories = wp.blocks.getCategories();
+		this.nodes = {};
 		this.state = {
-			filterValue: ''
+			filterValue: '',
+			focusedElementRef: null
 		};
 		this.filter = this.filter.bind( this );
+		this.isShownBlock = this.isShownBlock.bind( this );
+		this.changeMenuSelection = this.changeMenuSelection.bind( this );
+		this.setSearchFocus = this.setSearchFocus.bind( this );
+		this.bindReferenceNode = this.bindReferenceNode.bind( this );
+	}
+
+	bindReferenceNode( nodeName ) {
+		const binder = ( node ) => this.nodes[ nodeName ] = node;
+		binder.bind( this );
+
+		return binder;
+	}
+
+	componentDidMount() {
+		// Build a stream of keydown events on component mount.
+		this.keyStream$ = Observable.fromEvent( document, 'keydown' );
+
+		// Out of keydown stream build a stream matching arrow down and arrow right.
+		this.focusNextFocusableRef$ = this.keyStream$.filter( keydown => keydown.code === 'ArrowDown' || keydown.code === 'ArrowRight' || ( keydown.code === 'Tab' && keydown.shiftKey === false ) );
+
+		// Out of keydown stream build a stream matching arrow left and arrow up.
+		this.focusPreviousFocusableRef$ = this.keyStream$.filter( keydown => keydown.code === 'ArrowUp' || keydown.code === 'ArrowLeft' || ( keydown.code === 'Tab' && keydown.shiftKey === true ) );
+
+		// Get escape key presses.
+		this.closeInserter$ = this.keyStream$.filter( keydown => keydown.code === 'Escape' );
+
+		// Create subscriptions for our next and previous blockType focus actions.
+		this.nextBlockSubscription = this.focusNextFocusableRef$.subscribe( ( keydown ) => {
+			keydown.preventDefault();
+			focusNextFocusableRef( this );
+		} );
+		this.previousBlockSubscription = this.focusPreviousFocusableRef$.subscribe( ( keydown ) => {
+			keydown.preventDefault();
+			focusPreviousFocusableRef( this );
+		} );
+		this.closeInserterSubscription = this.closeInserter$.subscribe( () => {
+			this.props.closeMenu();
+		} );
+	}
+
+	componentWillUnmount() {
+		// Unsubscribe from the stream.
+		this.nextBlockSubscription.unsubscribe();
+		this.previousBlockSubscription.unsubscribe();
+		this.closeInserterSubscription.unsubscribe();
+
+		// These deletes are most likely not necessary but memory leaks can be pretty ugly.
```

Comment:
I think these are necessary to avoid using `this` on the subscribe callback while the component has been unmounted.